base/strings/basic.jl

```diff
@@ -621,6 +621,29 @@ isascii(c::AbstractChar) = UInt32(c) < 0x80
     return 0 ≤ r < 0x80
 end
 
+
+# Julia Strings and Arrays are stored in the following way
+# | 8-Byte Type Tag|  8-Byte size   | N-Bytes of UInt8 Code units |
+# |This data is safe to read from   |<- String pointer is address of first codeunit byte
+@inline function _isascii_subqword(bytes::AbstractVector{UInt8},first,n)
+    # Loading bytes before the string is safe because the bytes in front of a string are the sure to be owned by the string object
```

Comment:
The out of bounds data is shifted out, so from an implementation/algorithm detail it should be fine.  The only problem I see is if this triggered some kernel/os level memory access violation type of error when reading outside memory owned by a given application.

Reply:
Observing padding is nevertheless UB - irrespective of kernel/OS. Additionally, `pointer(::AbstractVector)` does not guarantee that the memory before the pointer is valid, only that the pointer itself is. If the pointer lies at the start of a page, dereferencing a decremented pointer will lead to a segfault. This is not safe to do.

Reply:
This works as a replacement but there is a bunch of stuff in that native that i don't understand
```julia
@inline function _isascii_subqword2(bytes::AbstractVector{UInt8},first,n)
    # Loading bytes before the string is safe because the bytes in front of a string are the sure to be owned by the string object
    # This is important because those bytes are technically owned by the same object( which if you read past the end may not be the case)
    n = ifelse(n < 7, n, 7) 
    n_trash = 8 - n
    #It would be nice to use reinterpret here but there is no good way to get at a negative index
    q = Ref{UInt64}()
    GC.@preserve q  begin
        qptr = Ptr{UInt8}(Base.unsafe_convert(Ref{UInt64}, q))
        Base._memcpy!(qptr, pointer(bytes,1), n)
    end
    qword = q[] << (n_trash * 8)
    return iszero(qword & UInt64(0x8080808080808080))
end
```

Reply:
```
Base.@propagate_inbounds function _isascii_subqword3(bytes::V,first,n) where V <: AbstractVector{UInt8}
    n = min(n, 7)
    @boundscheck checkbounds(bytes, first:(first+n))
    z = zero(UInt64)
    for i in 1:n
        z |= @inbounds bytes[first+i] << (i-1)
    end
    return iszero(z & 0x8080808080808080)
end
``` 

This benchmarks at the exact same speed as your `_isascii_subqword2` (which is faster than your original `_isascii_subqword` for me), but without pointer shenanigans. That may be due to AVX512, as I see things like [`vmovdqu8`](https://www.felixcloutier.com/x86/movdqu:vmovdqu8:vmovdqu16:vmovdqu32:vmovdqu64) in my assembly, which is new in AVX512.

Generally though, on non-AVX512 zero extension is a bit [sad](https://www.felixcloutier.com/x86/pmovzx).